ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixClusterManager.java

```diff
@@ -0,0 +1,563 @@
+        dcZkInfo.helixManager.addConfigChangeListener(dcZkInfo.clusterChangeListener);
+        dcZkInfo.helixManager.addLiveInstanceChangeListener(dcZkInfo.clusterChangeListener);
+        dcZkInfo.clusterChangeListener.waitForInitialization();
+      }
+    } catch (Exception e) {
+      helixClusterManagerMetrics.initializeInstantiationMetric(false);
+      throw new IOException("Encountered exception parsing json or connecting to Helix", e);
+    }
+    helixClusterManagerMetrics.initializeInstantiationMetric(true);
+    helixClusterManagerMetrics.initializeDatacenterMetrics();
+    helixClusterManagerMetrics.initializeDataNodeMetrics();
+    helixClusterManagerMetrics.initializeDiskMetrics();
+    helixClusterManagerMetrics.initializePartitionMetrics();
+    helixClusterManagerMetrics.initializeCapacityMetrics();
+  }
+
+  /**
+   * Populate the initial data from the admin connection. Create nodes, disks, partitions and replicas for the entire
+   * cluster.
+   * @throws Exception if creation of {@link AmbryDataNode}s or {@link AmbryDisk}s throw an Exception.
+   */
+  private void initialize() throws Exception {
+    for (DcZkInfo dcZkInfo : dcToDcZkInfo.values()) {
+      HelixAdmin admin = dcZkInfo.helixManager.getClusterManagmentTool();
+      for (String instanceName : admin.getInstancesInCluster(clusterName)) {
+        InstanceConfig instanceConfig = admin.getInstanceConfig(clusterName, instanceName);
+        AmbryDataNode datanode =
+            new AmbryDataNode(dcZkInfo.dcName, clusterMapConfig, instanceConfig.getHostName(), instanceConfig.getPort(),
+                getRackIdStr(instanceConfig), getSslPortStr(instanceConfig));
+        initializeDisksAndReplicasOnNode(datanode, instanceConfig);
+        instanceNameToDataNodeIdMap.put(instanceName, datanode);
+        dcZkInfo.clusterChangeListener.allInstances.add(instanceName);
+      }
+    }
+    for (Set<AmbryDisk> disks : dataNodeIdToDiskIds.values()) {
+      for (AmbryDisk disk : disks) {
+        clusterWideRawCapacityBytes += disk.getRawCapacityInBytes();
+      }
+    }
+    for (Map.Entry<AmbryPartition, Set<AmbryReplica>> partitionInfo : partitionIdToReplicaIds.entrySet()) {
+      long replicaCapacity = partitionInfo.getValue().iterator().next().getCapacityInBytes();
+      clusterWideAllocatedRawCapacityBytes += replicaCapacity * partitionInfo.getValue().size();
+      clusterWiseAllocatedUsableCapacityBytes += replicaCapacity;
+    }
+  }
+
+  /**
+   * Initialize the disks and replicas on the given node. Create partitions if this is the first time a replica of
+   * that partition is being constructed.
+   * @param datanode the {@link AmbryDataNode} that is being initialized.
+   * @param instanceConfig the {@link InstanceConfig} associated with this datanode.
+   * @throws Exception if creation of {@link AmbryDisk} throws an Exception.
+   */
+  private void initializeDisksAndReplicasOnNode(AmbryDataNode datanode, InstanceConfig instanceConfig)
```

Comment:
It's true that all the information can be gathered from the code but it would be nice to have it documented what an `InstanceConfig` contains (or is expected to contain) for easy reference.